phpBB/phpbb/report/controller/report.php

```diff
@@ -0,0 +1,268 @@
+	 * @throws \phpbb\exception\http_exception when $mode or $id is invalid for some reason
+	 */
+	public function handle($mode, $id)
+	{
+		// Get report handler
+		$this->report_handler = $this->report_handler->get($mode);
+
+		$this->user->add_lang('mcp');
+
+		// Set up variables
+		$this->tmp_vars = array(
+			'mode'		=> $mode,
+			'id'		=> $id,
+			'form'		=> array(
+				'error'				=> array(),
+				's_hidden_fields'	=> '',
+				'report_text'		=> $this->request->variable('report_text', '', true),
+				'reason_id'			=> $this->request->variable('reason_id', 0),
+				'user_notify'		=> ($this->user->data['is_registered']) ? $this->request->variable('notify', 0) : false,
+				'submit'			=> $this->request->variable('submit', ''),
+				'cancel'			=> $this->request->variable('cancel', ''),
+			),
+		);
+
+		$redirect_url = append_sid(
+			$this->phpbb_root_path . ( ($mode === 'pm') ? 'ucp' : 'viewtopic' ) . ".{$this->php_ext}",
+			($mode == 'pm') ? "i=pm&mode=view&p=$id" : "p=$id"
+		);
+		$redirect_url .= ($mode === 'post') ? "#p$id" : '';
+
+		// Set up CAPTCHA
+		if ($this->config['enable_post_confirm'] && !$this->user->data['is_registered'])
+		{
+			$this->tmp_vars['captcha'] = $this->captcha_factory->get_instance($this->config['captcha_plugin']);
+			$this->tmp_vars['captcha']->init(CONFIRM_REPORT);
+		}
+
+		//Has the report been cancelled?
+		if (!empty($this->tmp_vars['form']['cancel']))
+		{
+			redirect($redirect_url);
+		}
+
+		// Check CAPTCHA
+		if (!empty($this->tmp_vars['form']['submit']))
+		{
+			if (isset($this->tmp_vars['captcha']))
+			{
+				$visual_confirmation_response = $this->tmp_vars['captcha']->validate();
+				if ($visual_confirmation_response)
+				{
+					$this->tmp_vars['form']['error'][] = $visual_confirmation_response;
+				}
+			}
+
+			// Handle CAPTCHA
+			if (isset($this->tmp_vars['captcha']))
+			{
+				if (sizeof($this->tmp_vars['form']['error']) === 0)
+				{
+					$this->tmp_vars['captcha']->reset();
+				}
+				else if ($this->tmp_vars['captcha']->is_solved() !== false)
+				{
+					$this->tmp_vars['form']['s_hidden_fields'] .= build_hidden_fields(
+						$this->tmp_vars['captcha']->get_hidden_fields()
+					);
+				}
+			}
+		}
+
+		try
+		{
+			if (!empty($this->tmp_vars['form']['submit']) && sizeof($this->tmp_vars['form']['error']) === 0)
+			{
+				$this->report_handler->add_report(
+					(int) $id,
+					(int) $this->tmp_vars['form']['reason_id'],
+					(string) $this->tmp_vars['form']['report_text'],
+					(int) $this->tmp_vars['form']['user_notify']
+				);
+
+				// Send success message
+				switch ($mode)
+				{
+					case 'pm':
+						$lang_return = $this->user->lang['RETURN_PM'];
+						$lang_success = $this->user->lang['PM_REPORTED_SUCCESS'];
+						break;
+					case 'post':
+						$lang_return = $this->user->lang['RETURN_TOPIC'];
+						$lang_success = $this->user->lang['POST_REPORTED_SUCCESS'];
+						break;
+				}
+
+				meta_refresh(3, $redirect_url);
```

Comment:
@Nicofuma I don't see how that could be done nicely, since all meta_refresh() does is assigning template variables, which is only possible if we add a new service (we need to display a message). maybe a method for that in controller/helper would be a cleanest way to go.